storage/SMAIndexSubBlock.cpp

```diff
@@ -0,0 +1,665 @@
+                           const TypedValue &min,
+                           const TypedValue &max,
+                           const UncheckedComparator *less_comparator,
+                           const UncheckedComparator *equals_comparator) {
+  switch (comparison) {
+    case ComparisonID::kEqual:
+      return getSelectivity_E(literal, min, max, less_comparator);
+    case ComparisonID::kLess:
+      return getSelectivity_L(literal, min, max, less_comparator, equals_comparator);
+    case ComparisonID::kLessOrEqual:
+      return getSelectivity_LE(literal, min, max, less_comparator, equals_comparator);
+    case ComparisonID::kGreater:
+      return getSelectivity_G(literal, min, max, less_comparator, equals_comparator);
+    case ComparisonID::kGreaterOrEqual:
+      return getSelectivity_GE(literal, min, max, less_comparator, equals_comparator);
+    default:
+      return Selectivity::kUnknown;
+  }
+}
+
+SMAPredicate* SMAPredicate::ExtractSMAPredicate(const ComparisonPredicate &predicate) {
+  if (predicate.getLeftOperand().hasStaticValue()) {
+    DLOG_IF(FATAL, predicate.getRightOperand().getDataSource() != Scalar::kAttribute);
+    return new SMAPredicate(
+         static_cast<const ScalarAttribute&>(predicate.getRightOperand()).getAttribute().getID(),
+         flipComparisonID(predicate.getComparison().getComparisonID()),
+         predicate.getLeftOperand().getStaticValue().makeReferenceToThis());
+  } else {
+    DLOG_IF(FATAL, predicate.getLeftOperand().getDataSource() != Scalar::kAttribute);
+    return new SMAPredicate(
+        static_cast<const ScalarAttribute&>(predicate.getLeftOperand()).getAttribute().getID(),
+        predicate.getComparison().getComparisonID(),
+        predicate.getRightOperand().getStaticValue().makeReferenceToThis());
+  }
+}
+
+/**
+ * @brief Summation will promote lower precision types to higher precision types.
+ *
+ * @param type The type id of the type we wish to sum over.
+ * @return The higher precision sum typeid. Returns kNullType to indicate that the
+ *         given type cannot be summed (in lieu of a better sigil).
+ */
+inline TypeID sumType(TypeID type) {
+  switch (type) {
+    case kInt:
+    case kLong:
+      return kLong;
+    case kFloat:
+    case kDouble:
+      return kDouble;
+    default: {
+      return kNullType;
+    }
+  }
+}
+
+/**
+ * @param type A type id.
+ * @return True if the type can be summed.
+ */
+inline bool canSum(TypeID type) {
+  return sumType(type) != kNullType;
+}
+
+/**
+ * @brief Sets the given entry's sum to zero using the correct promoted type
+ *        of the sum. Does nothing if the entry's type cannot be summed.
+ * @note We ignore clearing any old data held in the previous sum because Double
+ *       and Long types are always represented inline.
+ *       
+ * @param entry A pointer to the entry to modify.
+ */
+inline void setTypedValueForSum(SMAEntry *entry) {
+  TypeID sum_type = sumType(entry->type_id);
```

Comment:
It seems like letting it crash would take almost as much time to catch the bug. As in, the process would be Run program -> Seg Fault -> Run it through gdb and find the dereference.

Versus Run Program -> Crash, read error message.

There's a lot of pointer referencing going on in this class, so I'm going hesitate adding DCHECKs unless there is a reason to add a check in certain dereferences and not others.